### PR TITLE
Fix panic in `struct` function with mixed scalar/array arguments

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Fixes https://github.com/apache/arrow-datafusion/issues/9773

## Rationale for this change
Bug (see https://github.com/apache/arrow-datafusion/issues/9773)

## What changes are included in this PR?

Fix bug by calling the correct api, add test

## Are these changes tested?
Yes, new test

## Are there any user-facing changes?
Less bugs

### CODE REVIEW

File: datafusion/functions/src/core/struct.rs [hunk 1]
```diff
@@ -47,17 +47,10 @@ fn array_struct(args: &[ArrayRef]) -> Result<ArrayRef> {
...
 }
+
 /// put values in a struct array.
 fn struct_expr(args: &[ColumnarValue]) -> Result<ColumnarValue> {
-    let arrays = args
-        .iter()
-        .map(|x| {
-            Ok(match x {
-                ColumnarValue::Array(array) => array.clone(),
-                ColumnarValue::Scalar(scalar) => scalar.to_array()?.clone(),
```
Comment (alamb) [inline datafusion/functions/src/core/struct.rs:57]: the bug here is that this always makes a 1 row array, even if one of the other arguments has more than 1 row.

**Review by yyy1000 — APPROVED** [review]
I reviewed it and it looks good to me :)
(when I port this function I didn't notice this XD

**Comment by alamb** [comment]
> I reviewed it and it looks good to me :) (when I port this function I didn't notice this XD

Thanks @yyy1000  ! 

It certainly didn't have test coverage either, and I think that bug has existed for quite a while

**Review by viirya — APPROVED** [review]

**Comment by alamb** [comment]
Thank you for the review @viirya

**Pull request merged**